plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -819,6 +849,99 @@ private void finishOptimize(ConnectorSession session, IcebergTableExecuteHandle
+
+        Table table = catalog.loadTable(session, executeHandle.getSchemaTableName());
+        Map<String, String> properties = table.properties();
+        if (properties.containsKey(WRITE_LOCATION_PROVIDER_IMPL)) {
+            throw new TrinoException(NOT_SUPPORTED, "Table " + executeHandle.getSchemaTableName() + " specifies " + properties.get(WRITE_LOCATION_PROVIDER_IMPL) +
+                    " as a location provider. Writing to Iceberg tables with custom location provider is not supported.");
+        }
+        Duration retention = requireNonNull(vacuumHandle.getRetentionThreshold(), "retention is null");
+        Duration minRetention = requireNonNull(getVacuumMinRetention(session), "vacuumMinRetention is null");
+        checkArgument(retention.compareTo(
+                minRetention) >= 0,
+                "Retention specified (%s) is shorter than the minimum retention configured in the system (%s). " +
+                        "Minimum retention can be changed with %s configuration property or iceberg.%s session property",
+                retention,
+                minRetention,
+                IcebergConfig.VACUUM_MIN_RETENTION,
+                IcebergSessionProperties.VACUUM_MIN_RETENTION);
+
+        long expireTimestampMillis = System.currentTimeMillis() - retention.toMillis();
+        table.expireSnapshots()
+                .expireOlderThan(expireTimestampMillis)
+                .cleanExpiredFiles(true)
+                .commit();
+        deleteOrphanFiles(table, session, executeHandle.getSchemaTableName(), expireTimestampMillis);
+        deleteOrphanMetadataFiles(table, session, executeHandle.getSchemaTableName(), expireTimestampMillis);
+    }
+
+    private void deleteOrphanFiles(Table table, ConnectorSession session, SchemaTableName schemaTableName, long expireTimestamp)
+    {
+        Set<String> validDataFilePaths = Streams.stream(table.snapshots())
+                .map(Snapshot::snapshotId)
+                .flatMap(snapshotId -> Streams.stream(table.newScan().useSnapshot(snapshotId).planFiles()))
+                .map(fileScanTask -> URI.create(fileScanTask.file().path().toString()).getPath())
+                .collect(toImmutableSet());
+        scanAndDeleteNotValidFiles(table, session, schemaTableName, expireTimestamp, validDataFilePaths, "/data/");
+    }
+
+    private void deleteOrphanMetadataFiles(Table table, ConnectorSession session, SchemaTableName schemaTableName, long expireTimestamp)
+    {
+        ImmutableSet<String> manifests = Streams.stream(table.snapshots())
+                .flatMap(snapshot -> snapshot.allManifests().stream())
+                .map(ManifestFile::path)
+                .collect(toImmutableSet());
+        List<String> manifestLists = ReachableFileUtil.manifestListLocations(table);
+        List<String> otherMetadataFiles = Streams.concat(
+                ReachableFileUtil.metadataFileLocations(table, false).stream(),
```

Comment:
is `false` ok here. I do not fully undestand the difference

Reply:
I pretty much copied it from iceberg spark vacuum procedure ;)